### PR TITLE
Handle deprecation attribute for definition-like constructions

### LINKED ISSUE #12266: "deprecated" attribute for Lemmas.
Feature Request: It would be nice to be able to mark some lemmas as "deprecated" so compilation warning will be issued when they are used. There’s the deprecated attribute 3, but it doesn’t work for Lemmas.

There are workarounds (via Notation) but they look rather clumsy to me.
Related discussion: https://coq.discourse.group/t/deprecated/814

### LINKED ISSUE #17531: Incorrect glob with abbreviation in pattern
~~~coq
Notation bar := option.

Definition foo (x: nat) : nat :=
  match x with
  | 0 => 0
  | S bar => bar
  end.
~~~
The `bar` in `S bar` is reported in the .glob as being the Notation.

### PR DESCRIPTION
**Pull request by maximedenes**
Closes #12266

Also close https://github.com/coq/coq/issues/17531 

Overlays:
- https://github.com/LPCIC/coq-elpi/pull/456
- https://github.com/mattam82/Coq-Equations/pull/546
- https://github.com/Mtac2/Mtac2/pull/379
- https://github.com/MetaCoq/metacoq/pull/959

### CODE REVIEW

File: interp/constrintern.ml [hunk 1]
```diff
@@ -1260,14 +1260,26 @@ let glob_sort_of_level (level: glob_level) : glob_sort =
...
   let loc = qid.loc in
   match intern_extended_global_of_qualid qid with
   | TrueGlobal (GlobRef.VarRef _) when no_secvar ->
       (* Rule out section vars since these should have been found by intern_var *)
       raise Not_found
-  | TrueGlobal ref -> (DAst.make ?loc @@ GRef (ref, us)), Some ref, args
+  | TrueGlobal ref ->
+    begin match ref with
+    | GlobRef.ConstRef c ->
+      Option.iter (fun d -> warn_deprecated_constant (ref,d)) (Deprecation.lookup_constant_deprecation c);
```
Comment (SkySkimmer) [inline interp/constrintern.ml:1279]: I wonder if this is the right place to warn
ie if I do `Existing Instance some_deprecated_constant` it won't go through this code.

Comment (maximedenes) [reply interp/constrintern.ml:1279]: Right, there's this and a few other ways we can depend on a deprecated constant and escape the warnings (if it has been inserted by unification, for example).

So where would you put the warning emitter?

Comment (SkySkimmer) [reply interp/constrintern.ml:1279]: Probably Nametab.locate_extended
probably merge the syndef warning with it

**Comment by andres-erbsen** [comment]
I looked into writing a script to automatically deprecate definitions changed by a diff and this feature would be really helpful. @maximedenes do you intend to return to this PR? If not, I might try to figure out how to incorporate the suggestion by @SkySkimmer (though I am not familiar with the relevant files).

**Comment by maximedenes** [comment]
@andres-erbsen I plan to return to it at some point, but not immediately, so feel free to push commits here.

**Comment by SkySkimmer** [comment]
Rebased and now emitting the warning from nametab

@coqbot run full ci

**Comment by ppedrot** [comment]
Shouldn't the deprecation module go to library rather than lib?

**Comment by SkySkimmer** [comment]
I see it as an extension of cwarning which is in lib

**Comment by ppedrot** [comment]
It's about naming stuff so it should live where the nametab is. Lib is way too low in my opinion.

**Comment by SkySkimmer** [comment]
I don't think deprecation is about naming stuff. We could very well want to emit a deprecation warning in engine/ or some such thing.

**Comment by Alizter** [comment]
@SkySkimmer Can you update the refman to mention the attribute in the definition-like constructions?

**Comment by SkySkimmer** [comment]
done

File: doc/sphinx/using/libraries/writing.rst [hunk 2]
```diff
@@ -9,8 +9,8 @@ available in the Coq repository in `doc/plugin_tutorial
 Deprecating library objects or tactics
 --------------------------------------
 
-You may use the following :term:`attribute` to deprecate a notation or
-tactic.  When renaming a definition or theorem, you can introduce a
+You may use the following :term:`attribute` to deprecate a notation,
+tactic, definition, axiom or theorem.  When renaming a definition or theorem, you can introduce a
 deprecated compatibility alias using :cmd:`Notation (abbreviation)`
 (see :ref:`the example below <compatibility-alias>`).
```
Comment (Alizter) [inline doc/sphinx/using/libraries/writing.rst:15]: This workaround is probably no longer relevant.

**Review by Alizter — APPROVED** [review]

**Comment by Alizter** [comment]
I will not request another CI since there was a successful run before the doc push.

**Comment by Alizter** [comment]
Let's check the doc build then I will merge.

File: doc/changelog/08-vernac-commands-and-options/16890-deprecate-definitions.rst [hunk 3]
```diff
@@ -0,0 +1,5 @@
+- **Added:**
+  The `deprecated` attribute is now supported for definition-like constructions
```
Comment (jfehrle) [inline doc/changelog/08-vernac-commands-and-options/16890-deprecate-definitions.rst:2]:
```suggestion
  The :attr:`deprecated` attribute is now supported for definition-like constructions
```

**Comment by Alizter** [comment]
@coqbot merge now

**Pull request merged**